Add individual simulations plots to summary plot.

verveerpj opened the pull request:
Added the option to plot individual simulations to the summary plot, much like the objective function plot was doing.

It should be fairly easy to generalize the SummaryPlot to a plugin that can  handle the objective functions also.

oyvindeide commented on everviz/pages/summary_values.py:
```diff
@@ -65,34 +63,38 @@ def _set_up_data_sources(api, keys=None):
 
 def page_layout(api):
     sources = _set_up_data_sources(api)
-    if sources is None:
```
Can sources no longer be None?

verveerpj replied:
That would be the case that an optimization does not use eclipse or flow and does not produce summary values.

oyvindeide replied:
So we need that check, running the plugin on a case without summary values gives `AttributeError`.

Can you also add a test simulating that condition?

verveerpj replied:
done.

oyvindeide commented:
Could you also add an integration test of the whole plugin? That would make it much easier to review without having a case with summary values available, and seeing as there is quite a bit of logic in the plugin, that would also be beneficial.

verveerpj commented:
> Could you also add an integration test of the whole plugin? That would make it much easier to review without having a case with summary values available, and seeing as there is quite a bit of logic in the plugin, that would also be beneficial.

Basically like the current crossplot integration test?

oyvindeide commented:
> Basically like the current crossplot integration test?

Yes

verveerpj commented:
I have now added an integration test, and fixed one unit test.

Regarding the integration test: is there a way to select a radio button using `dash_duo`, similar to `dash_duo.select_dcc_dropdown`? I have not been able to find any documentation or hints about that anywhere. It would be nice to be able to test selecting a radio button.

oyvindeide commented:
> I have now added an integration test, and fixed one unit test.
> 
> Regarding the integration test: is there a way to select a radio button using `dash_duo`, similar to `dash_duo.select_dcc_dropdown`? I have not been able to find any documentation or hints about that anywhere. It would be nice to be able to test selecting a radio button.

Great, thanks! Will have a look later today or tomorrow. Yes, there is a way, but not as pretty, can do this:

```
    text = dash_duo.find_element(
        "#{} label:nth-child({})".format("radio_id", <index>)
    ).text
```
That gives you the text of the radio button to verify against, or you could select something else.

verveerpj commented:
> Great, thanks! Will have a look later today or tomorrow. Yes, there is a way, but not as pretty, can do this:
> 
> ```
>     text = dash_duo.find_element(
>         "#{} label:nth-child({})".format("radio_id", <index>)
>     ).text
> ```
> 
> That gives you the text of the radio button to verify against, or you could select something else.

That gives access to the label text, right? Can I also select, or click?

oyvindeide commented:
> 
> That gives access to the label text, right? Can I also select, or click?

Haven´t tried, but think so yes, just returning the element I think you should be able to do: `element.click()`

verveerpj commented:
> > That gives access to the label text, right? Can I also select, or click?
> 
> Haven´t tried, but think so yes, just returning the element I think you should be able to do: `element.click()`

no, that does not work, it complains it is  not a clickable element. So I guess the string argument to find_element needs to be something different.

oyvindeide commented:
> no, that does not work, it complains it is not a clickable element. So I guess the string argument to find_element needs to be something different.

Weird, works on my machine, did: 
```
    element = dash_duo.find_element(
        "#{} label:nth-child({})".format("radio_id", <index>)
    )
    element.click()
```
Note however that it´s 1 indexed, could that be it? My example is from: `test_crossplot_layout_radio` in `test_layout.py`.

verveerpj commented:
> Note however that it´s 1 indexed, could that be it? My example is from: `test_crossplot_layout_radio` in `test_layout.py`.

No, it turned out to be a quite different issue!

In my code, just before doing this, I had cleared a dropdown for the x-axis. When you do that, it opens that dropdown, and it stays open. That overlapped with the radio buttons, which are just below it. So, when you simulate the click on the radio buttion, the click actually goes to the dropdown, and not the radio buttons... This is detected, hence the error that the click goes to another element.

Moving the clearing of the dropdown to appear after this, fixed it.

oyvindeide commented on tests/integration/crossplot/test_summary_plot_integration.py:
```diff
@@ -0,0 +1,63 @@
+from datetime import datetime
```
Thanks for adding this test, made my review a lot easier 👍 Agree with the test, but perhaps put it in a different folder?

verveerpj replied:
oh, guess I forgot to move that :smile:, will do...

verveerpj commented on tests/unit/test_summary_plot.py:
```diff
@@ -3,53 +3,81 @@
+    assert layout == {}
+
+
 def test_summary_values():
```
You are correct, this test function is too long. I have split it in three logically distinct test functions. They need to remain unit tests though, they test one specific function.

verveerpj commented on everviz/pages/summary_values.py:
```diff
@@ -8,38 +8,32 @@
 logger = get_logger()
 
 
 def _summary_values(summary_values):
```
Indeed, I have split it up and it makes the code clearer.

oyvindeide commented on tests/integration/summary_plot/test_summary_plot_integration.py:
```diff
@@ -0,0 +1,62 @@
+from datetime import datetime
```
We both omitted having an `__init__.py` in these folders, me in `integration` and you here, could you add them both? That gives some `pylint` errors, other than that this looks good now 👍
